File: lingting-spring-redis/src/main/java/live/lingting/spring/redis/prefix/JdkKeyPrefixSerializer.java
package live.lingting.spring.redis.prefix;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.serializer.JdkSerializationRedisSerializer;

/**
 * @author lingting 2024-04-17 16:10
 */
@RequiredArgsConstructor
public class JdkKeyPrefixSerializer extends JdkSerializationRedisSerializer {

	private final KeyPrefixConvert convert;

	@Override
	public byte[] serialize(Object value) {
		byte[] bytes = super.serialize(value);
		return convert.isMatch(bytes) ? convert.wrap(bytes) : bytes;
	}

	@Override
	public Object deserialize(byte[] bytes) {
		byte[] target = convert.isMatch(bytes) ? convert.unwrap(bytes) : bytes;
		return super.deserialize(target);
	}

}

File: lingting-spring-redis/src/main/kotlin/live/lingting/spring/redis/prefix/JdkKeyPrefixSerializer.kt
package live.lingting.spring.redis.prefix

import org.springframework.data.redis.serializer.JdkSerializationRedisSerializer

/**
 * @author lingting 2024-04-17 16:10
 */
@Suppress("WRONG_NULLABILITY_FOR_JAVA_OVERRIDE")
class JdkKeyPrefixSerializer(private val convert: KeyPrefixConvert) : JdkSerializationRedisSerializer() {

    override fun serialize(value: Any?): ByteArray {
        val bytes = super.serialize(value)
        return if (convert.isMatch(bytes)) convert.wrap(bytes) else bytes
    }

    override fun deserialize(bytes: ByteArray?): Any {
        val target = if (bytes != null && convert.isMatch(bytes)) convert.unwrap(bytes) else bytes
        return super.deserialize(target)
    }

}
